### PR TITLE
implemented shopping cart

### PR DESCRIPTION
**Pull request by volodymyrruso**

### CODE REVIEW

File: src/main/java/org/mate/bookstorespringboot/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+@RequestMapping("/api/cart")
+@RequiredArgsConstructor
+@Validated
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @GetMapping()
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/controller/ShoppingCartController.java]:
```suggestion
    @GetMapping
```

File: src/main/java/org/mate/bookstorespringboot/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+    public ResponseEntity<ShoppingCartDto> updateShoppingCart(Authentication authentication,
+                                                              @RequestBody @Validated
+                                                              UpdateCartItemRequestDto requestDto,
+                                                              @PathVariable Long cartItemId) {
+        return ResponseEntity.ok(shoppingCartService
+                .updateCartItemInShoppingCart(getUserIdByAuthentication(authentication),
+                        cartItemId, requestDto));
+    }
+
+    @DeleteMapping("/{cartItemId}")
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/controller/ShoppingCartController.java:54]: ResponseStatus NoContent

File: src/main/java/org/mate/bookstorespringboot/controller/dto/cartitem/CartItemRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,8 @@
+package org.mate.bookstorespringboot.controller.dto.cartitem;
+
+import jakarta.annotation.Nonnull;
+import jakarta.validation.constraints.Min;
+
+public record CartItemRequestDto(@Min(1) int quantity,
+                                 @Nonnull Long bookId) {
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/controller/dto/cartitem/CartItemRequestDto.java]: Better use annotation from validators package.

File: src/main/java/org/mate/bookstorespringboot/repository/CartItemRepository.java [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+import org.mate.bookstorespringboot.model.CartItem;
+import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface CartItemRepository extends JpaRepository<CartItem, Long> {
+
+    @Query("SELECT ci FROM CartItem ci "
+            + "LEFT JOIN FETCH ci.book b "
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/repository/CartItemRepository.java]: Why left join?

File: src/main/java/org/mate/bookstorespringboot/repository/ShoppingCartRepository.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface ShoppingCartRepository extends JpaRepository<ShoppingCart, Long> {
+    @Query("""
+            SELECT sc FROM ShoppingCart sc
+             LEFT JOIN FETCH sc.cartItem ci
+            LEFT JOIN FETCH sc.cartItem.book
+            WHERE sc.user.id = :userId"""
+    )
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/repository/ShoppingCartRepository.java]: the same question. Use the same code style for queries

File: src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+import org.mate.bookstorespringboot.service.interfaces.ShoppingCartService;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartIServiceImpl implements ShoppingCartService {
+
+    private static final String SHOPPING_CART_NOT_FOUND = "Shopping cart not found with id: ";
+    private static final String CART_ITEM_NOT_FOUND = "Cart item not found with id: ";
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java]: I don't prefer to move the exception message to constant. Also, you have unused variables here

File: src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        return shoppingCartMapper.toDto(cart);
+    }
+
+    @Override
+    @Transactional
+    public ShoppingCartDto addCartItemToShoppingCart(Long userId,
+                                                     CartItemRequestDto cartItem) {
+        ShoppingCart shoppingCartToBeSaved = getShoppingCartByUserId(userId);
+
+        Book book = bookRepository.findById(cartItem.bookId())
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java]: You can call this method from BookService

File: src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                .orElseThrow(() -> new EntityNotFoundException("Book not found with id: "
+                        + cartItem.bookId()));
+
+        cartItemRepository.findByShoppingCartIdAndBookId(shoppingCartToBeSaved.getId(),
+                        cartItem.bookId())
+                .ifPresent(cartItem1 -> {
+                    throw new EntityNotFoundException("Book with id: "
+                            + cartItem.bookId()
+                            + " already in shopping cart");
+                });
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java]: Do you really need to throw an exception in this case? Maybe increase CartItem quantity?

File: src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+                                                        UpdateCartItemRequestDto
+                                                                    updateCartItemRequestDto) {
+        ShoppingCart shoppingCart = getShoppingCartByUserId(userId);
+        CartItem cartItem =
+                cartItemRepository.findByShoppingCartIdAndBookId(cartItemId, shoppingCart.getId())
+                        .map(item -> {
+                            item.setQuantity(updateCartItemRequestDto.quantity());
+                            return item;
+                        })
+                        .orElseThrow(() -> new jakarta.persistence.EntityNotFoundException(
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java]: Wrong import

File: src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+                                String.format("No cart item with id: %d for user: %d",
+                                        cartItemId, userId)
+                        ));
+        cartItemRepository.save(cartItem);
+        return shoppingCartMapper.toDto(shoppingCartRepository.save(shoppingCart));
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java:85]: When you have more than one save operation, better to use the Transaction

File: src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    @Override
+    public void deleteCartItemFromShoppingCart(Long userId, Long cartItemId) {
+
+        ShoppingCart shoppingCart = getShoppingCartByUserId(userId);
+        CartItem cartItem = getCartItemById(cartItemId);
+
+        shoppingCart.getCartItem().remove(cartItem);
```
Comment (liliia-ponomarenko) [inline src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java:95]: Did you update the shopping cart?

File: src/main/resources/db/changelog/changeset_create_table/create_shopping_cart_table.sql [hunk 12]
```diff
@@ -0,0 +1,5 @@
+CREATE TABLE shopping_cart (
+                              id BIGINT AUTO_INCREMENT PRIMARY KEY,
+                              user_id BIGINT,
+                              CONSTRAINT fk_user FOREIGN KEY (user_id) REFERENCES users(id)
+);
```
Comment (liliia-ponomarenko) [inline src/main/resources/db/changelog/changeset_create_table/create_shopping_cart_table.sql]: One empty line at the end of the file ;)

**Review by liliia-ponomarenko — CHANGES_REQUESTED** [review]
Good job! Let's improve your solution ;)

File: src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java (hunk 11 above)
Comment (volodymyrruso) [reply src/main/java/org/mate/bookstorespringboot/service/impl/ShoppingCartIServiceImpl.java:95]: what do you mean by update? it works good and gets deleted from the table

File: src/main/java/org/mate/bookstorespringboot/controller/dto/cartitem/CartItemRequestDto.java (hunk 3 above)
Comment (volodymyrruso) [reply src/main/java/org/mate/bookstorespringboot/controller/dto/cartitem/CartItemRequestDto.java]: Sorry, don't quite understand what is the validators package?

File: src/main/java/org/mate/bookstorespringboot/repository/CartItemRepository.java (hunk 4 above)
Comment (volodymyrruso) [reply src/main/java/org/mate/bookstorespringboot/repository/CartItemRepository.java]: updated

**Review by liliia-ponomarenko — APPROVED** [review]
Great! 😉

**Pull request merged**